habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/UvSetup.java

```
@@ -0,0 +1,115 @@
+package org.technologybrewery.habushu;
+
+import com.vdurmont.semver4j.Semver;
+import com.vdurmont.semver4j.Semver.SemverType;
+
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.logging.Log;
+import org.technologybrewery.habushu.exec.UvCommandHelper;
+import org.technologybrewery.habushu.util.UvUtil;
+
+import java.io.File;
+import java.util.List;
+
+/**
+ * Subclass of PythonPackageAndDependencyManagerSetup. Ensures uv-related pre-requisite tools that Habushu leverages are installed and available on the
+ * developer's machine to support the same functionality across multiple Mojo implementations. These include:
+ * <ul>
+ * <li>uv (installed version must satisfy {@link UvUtil#UV_VERSION_REQUIREMENT})</li>
+ * </ul>
+ */
+public class UvSetup extends AbstractPythonPackageAndDependencyManagerSetup {
+   
+    public UvSetup(String pythonVersion, File baseDir, boolean rewriteLocalPathDepsInArchives, Log log) throws MojoExecutionException {
+        super(pythonVersion, baseDir, rewriteLocalPathDepsInArchives, log); 
+    }
+
+    /**
+     * Checks if the .python-version file exists. If it exists, it checks the listed python 
+     * version and updates the listed version to the pythonVersion, if necessary. If the .python-version
+     * file does not exisit, it throws a MojoExecutionException.
+     */
```

I: Unlike Poetry, `uv` users don't have to install the desired python version before setting up the virtual environment. For this reason, I decided that the best way to check of the correct python version will be used in `uv` generated virtual environments is to verify that the `.python-version` file is present in the package directory. If it is not present, habushu will raise an error message saying it must be present.

---

I'm wondering if we want to default to a version of Python instead.  Failing loudly is good for forcing user action, but at the same time, this type of behavior may be a turnoff for Python developers.  Thoughts?

---

Q: Would it be possible to create the `.python-version` file with the correct version if it doesn't exist?

---

I thought about that. If we were to initialize a `uv`- based project via `habushu`, it would make sense for `habushu` to create the `.python-version` file. However, if a user already has a `uv`- based project setup and they're adding `habushu` to the project, I'm afraid that if `habushu` were to automatically create the `.python-version` file for the user, it could potentially mess up their current virtual environment.

---

> if a user already has a uv- based project setup and they're adding habushu to the project, I'm afraid that if habushu were to automatically create the .python-version file for the user

If this were true, wouldn't the user already have a  .python-version file as part of their uv project?

---

No. To create a virtual environment with `uv`, users have a few options.

1. They can do nothing and the latest available version of python will be used within the virtual environment. 
2. They can run `uv --python <desired-version>` to create a virtual environment with the desired version of python.
3. They can have a `.python-version` file in the module and `uv` will use that file to create the virtual environment with the specified Python version. 

I'm forcing users to use the last option since it's the only option that allows for reproducible builds.

---

Okay, if the goal is reproducible builds and we need that file to make it happen, I think it's fair to have them create the file otherwise fail

---

Thoughts? ^

---

Could we interrogate the python version from the venv and create the file with that?

---

yes, I could do that. I'll incorporate that in.